python/GafferSceneTest/IECoreScenePreviewTest/CapturingRendererTest.py

```diff
@@ -165,5 +168,390 @@ def testDeformingObject( self ) :
+
+			for subName, subObject in procExpanded.items():
+
+				if subName.startswith( "/" ):
+					newName = objectName + subName
+				else:
+					newName = objectName + "/" + subName
+
+				newAttrs = parentAttrs.copy()
+				for a, val in subObject.capturedAttributes.items():
+					newAttrs[a] = val
+
+				newTransformTimes = subObject.capturedTransformSampleTimes
+				if newTransformTimes != parentTransformSampleTimes:
+					if parentTransformSampleTimes == []:
+						# Combining a static transform with an animated transform results in an animated transform
+						pass
+					elif newTransformTimes == []:
+						newTransformTimes = parentTransformSampleTimes
+					else:
+						raise IECore.Exception( "Incompatible transform sample times when expanding procedural at location '%s' : %s != %s" % ( newName, parentTransformSampleTimes, newTransformTimes ) )
+
+				newTransforms = subObject.capturedTransforms
+				if parentTransforms == []:
+					pass
+				elif newTransforms == []:
+					newTransforms = parentTransforms
+				else:
+					effectiveParentTransforms = parentTransforms
+					if len( effectiveParentTransforms ) != len( newTransforms ):
+						if len( effectiveParentTransforms ) == 1:
+							effectiveParentTransforms = [ parentTransforms[0] ] * len( newTransforms )
+						elif len( newTransforms ) == 1:
+							newTransforms = [ newTransforms[0] ] * len( effectiveParentTransforms )
+						else:
+							raise IECore.Exception( "Incompatible transform sample lengths when expanding procedural at location '%s' : %s != %s" % ( newName, len( parentTransforms ), len( newTransforms ) ) )
+
+					newTransforms = [
+						newTransforms[i] * effectiveParentTransforms[i]
+						for i in range( len( newTransforms ) )
+					]
+
+				# \todo - If we want to test links properly, that would need better support ... passing
+				# object references from `procRenderer` doesn't feel reliable ... it's probably reasonable
+				# to just pass object names for comparison purposes, but then we should probably prefix all
+				# the names in the list, the same way we do the new object name?
+				# I don't think we currently really support light linking inside procedurals much, so I haven't
+				# worried about it yet.
+
+				result[ newName ] = CapturingRendererTest.ExpandedCapture(
+					subObject.capturedSamples, subObject.capturedSampleTimes, newTransforms, newTransformTimes,
+					newAttrs, subObject.capturedLinks
+				)
+
+		return result
+
+	@staticmethod
+	def assertRendersMatch( capturingA, capturingB, expandProcedurals = False, ignoreLinks = False ):
+		expandA = CapturingRendererTest.expandCapturingRenderer( capturingA, expandProcedurals )
+		expandB = CapturingRendererTest.expandCapturingRenderer( capturingB, expandProcedurals )
+
+		if expandA.keys() != expandB.keys():
+			raise IECore.Exception( "Objects do not match : %s != %s" % ( list( expandA.keys() ), list( expandB.keys() ) ) )
```

Comment:
I think we should be using some sort of `self.assert*()` here - I think there's a distinction between a test failure and an error during testing, and raising is an error rather than a failure. Note that you can pass custom messages to many of the `assert()` methods.

Reply:
Oh, I see that this is a static method and is called from `SceneTestCase`, so we don't have a `self` available. In that case I think we can just raise `AssertionError` rather than `Exception` and it'll be treated as a failure rather than an error.